File: z_lc/15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        
        def next_uniq(i, bound):
            i += 1
            while i < bound and nums[i] == nums[i-1]:
                i += 1
            return i
        def prev_uniq(i, bound):
            i -= 1
            while i > bound and nums[i] == nums[i+1]:
                i -= 1
            return i
        res = []
        for i in range(n - 2):
            if nums[i] > 0:
                break            
            if i > 0 and nums[i] == nums[i-1]:
                continue
            j, k = i + 1, n - 1
            while j < k:
                s = nums[i] + nums[j] + nums[k]
                if s == 0:
                    res.append([nums[i], nums[j], nums[k]])
                    j = next_uniq(j, k)
                    k = prev_uniq(k, j)
                elif s < 0:
                    j = next_uniq(j, k)
                else:
                    k = prev_uniq(k, j)
        return res

File: z_lc/15-3sum/3sum.rs
impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let n = nums.len();
        if n < 3 {
            return vec![];
        }
        
        let mut A = nums;
        A.sort();
        
        let next_uniq = |mut i, bound| {
            while {i += 1; i < bound && A[i] == A[i-1]} {}
            i
        };
        let prev_uniq = |mut i, bound| {
            while {i -= 1; i > bound && A[i] == A[i+1]} {}
            i
        };
        
        let mut res = vec![];
        for i in 0..(n - 2) {
            if A[i] > 0 {
                break;
            }
            if i > 0 && A[i] == A[i - 1] {
                continue;
            }
            let (mut j, mut k) = (i + 1, n - 1);
            while j < k {
                match A[i] + A[j] + A[k] {
                    0 => {
                        res.push(vec![A[i], A[j], A[k]]);
                        j = next_uniq(j, k);
                        k = prev_uniq(k, j);
                    },
                    x if x < 0 => j = next_uniq(j, k),
                    _ => k = prev_uniq(k, j),
                }
            }
        }
        res
    }
}